chips/apollo3/src/iom.rs

```diff
@@ -605,13 +612,68 @@ impl<'a> Iom<'_> {
                 }
 
                 self.spi_read_buffer.replace(buf);
+
+                if self.read_len.get() > self.read_index.get() {
+                    let remaining_bytes = (self.read_len.get() - self.read_index.get()).min(32);
+                    if remaining_bytes > 4 {
+                        self.registers
+                            .fifothr
+                            .modify(FIFOTHR::FIFORTHR.val(remaining_bytes as u32));
+                    } else {
+                        self.registers.fifothr.modify(FIFOTHR::FIFORTHR.val(1));
+                    }
+                } else {
+                    self.registers.fifothr.modify(FIFOTHR::FIFORTHR.val(0));
+                }
             } else {
-                while self.registers.fifoptr.read(FIFOPTR::FIFO1SIZ) > 0 {
+                while self.registers.fifoptr.read(FIFOPTR::FIFO0SIZ) > 0 {
                     let _d = self.registers.fifopop.get().to_ne_bytes();
                 }
+
+                self.registers.fifothr.modify(FIFOTHR::FIFORTHR.val(0));
             }
 
-            if self.write_len.get() > 0 && self.write_index.get() >= self.write_len.get() {
+            // Write more data out
+            if self.write_index.get() < self.write_len.get() {
+                if let Some(write_buffer) = self.buffer.take() {
+                    let mut transfered_bytes = 0;
+
+                    while self.registers.fifoptr.read(FIFOPTR::FIFO0REM) > 0
+                        && self.registers.fifoptr.read(FIFOPTR::FIFO1REM) > 4
+                        && self.write_index.get() < self.write_len.get()
+                        && transfered_bytes < 24
```

Comment:
Similar question here — what's up with the 4 byte threshold; also where did `24` come from?

Reply:
The 4 bytes here is to make sure that the read FIFO (reading from the SPI bus to the hardware FIFO) has at least 4 bytes free. Otherwise if we write 4 bytes we might overflow the FIFO. It really could be `>=` but `>` gives us a little wiggle room as the hardware does seem a little slow at updating the FIFO size registers.

The 24 is along the same lines, of just making sure we don't write too much data. I don't have a good answer of why it should be 24, but that seems to work reliably.